libbpf-rs/src/xdp.rs

```diff
@@ -0,0 +1,108 @@
+
+bitflags! {
+    /// Flags to configure the `XDP` operations
+    pub struct XdpFlags: u32 {
+        /// See [`libbpf_sys::XDP_FLAGS_UPDATE_IF_NOEXIST`].
+        const UPDATE_IF_NOEXIST = libbpf_sys::XDP_FLAGS_UPDATE_IF_NOEXIST as _;
+        /// See [`libbpf_sys::XDP_FLAGS_SKB_MODE`].
+        const SKB_MODE          = libbpf_sys::XDP_FLAGS_SKB_MODE as _;
+        /// See [`libbpf_sys::XDP_FLAGS_DRV_MODE`].
+        const DRV_MODE          = libbpf_sys::XDP_FLAGS_DRV_MODE as _;
+        /// See [`libbpf_sys::XDP_FLAGS_HW_MODE`].
+        const HW_MODE           = libbpf_sys::XDP_FLAGS_HW_MODE as _;
+        /// See [`libbpf_sys::XDP_FLAGS_REPLACE`].
+        const REPLACE           = libbpf_sys::XDP_FLAGS_REPLACE as _;
+        /// See [`libbpf_sys::XDP_FLAGS_MODES`].
+        const MODES             = libbpf_sys::XDP_FLAGS_MODES as _;
+        /// See [`libbpf_sys::XDP_FLAGS_MASK`].
+        const MASK              = libbpf_sys::XDP_FLAGS_MASK as _;
+    }
+
+}
+
+/// Represents a XDP program.
+///
+/// This struct exposes operations to attach, detach and query a XDP program
+#[derive(Clone, Copy, Debug)]
+pub struct Xdp<'fd> {
+    fd: BorrowedFd<'fd>,
+    attach_opts: libbpf_sys::bpf_xdp_attach_opts,
+    query_opts: libbpf_sys::bpf_xdp_query_opts,
+}
+
+impl<'fd> Xdp<'fd> {
+    /// Create a new xdp instance with the given file descriptor of the `SEC("xdp")` [`Program`][crate::Program].
+    pub fn new(fd: BorrowedFd<'fd>) -> Self {
+        let mut xdp = Xdp {
+            fd,
+            attach_opts: libbpf_sys::bpf_xdp_attach_opts::default(),
+            query_opts: libbpf_sys::bpf_xdp_query_opts::default(),
+        };
+        xdp.attach_opts.sz = size_of::<libbpf_sys::bpf_xdp_attach_opts>() as libbpf_sys::size_t;
+        xdp.query_opts.sz = size_of::<libbpf_sys::bpf_xdp_query_opts>() as libbpf_sys::size_t;
+        xdp
+    }
+
+    /// Attach the XDP program to the given interface to start processing the packets
+    pub fn attach(&mut self, ifindex: i32, flags: XdpFlags) -> Result<()> {
+        let ret = unsafe {
+            libbpf_sys::bpf_xdp_attach(
+                ifindex,
+                self.fd.as_raw_fd(),
+                flags.bits(),
+                &self.attach_opts as *const _,
+            )
+        };
+        util::parse_ret(ret)
+    }
+
+    /// Detach the XDP program from the interface
+    pub fn detach(&mut self, ifindex: i32) -> Result<()> {
+        let ret = unsafe {
+            libbpf_sys::bpf_xdp_detach(
+                ifindex,
+                XdpFlags::UPDATE_IF_NOEXIST.bits(),
+                &self.attach_opts as *const _,
+            )
+        };
+        util::parse_ret(ret)
+    }
+
+    /// Query to inspect the program
+    pub fn query(&mut self, ifindex: i32, flags: XdpFlags) -> Result<bpf_xdp_query_opts> {
+        let mut opts = self.query_opts;
+        let err = unsafe {
+            libbpf_sys::bpf_xdp_query(ifindex, flags.bits() as i32, (&mut opts) as *mut _)
```

Comment:
Why is this cast necessary?
```suggestion
            libbpf_sys::bpf_xdp_query(ifindex, flags.bits() as i32, &mut opts)
```